src/h3lib/lib/polyfill.c

```diff
@@ -406,23 +407,78 @@ void iterStepPolygonCompact(IterCellsPolygonCompact *iter) {
         iter->_started = true;
     }
 
+    ContainmentMode mode = FLAG_GET_CONTAINMENT_MODE(iter->_flags);
+
     while (cell) {
         int cellRes = H3_GET_RESOLUTION(cell);
 
         // Target res: Do a fine-grained check
         if (cellRes == iter->_res) {
-            // Check if the cell is in the polygon
-            // TODO: Handle other polyfill modes here
-            LatLng center;
-            H3Error centerErr = H3_EXPORT(cellToLatLng)(cell, &center);
-            if (NEVER(centerErr != E_SUCCESS)) {
-                iterErrorPolygonCompact(iter, centerErr);
-                return;
+            if (mode == CONTAINMENT_CENTER || mode == CONTAINMENT_OVERLAPPING) {
+                // Check if the cell center is inside the polygon
+                LatLng center;
+                H3Error centerErr = H3_EXPORT(cellToLatLng)(cell, &center);
+                if (centerErr != E_SUCCESS) {
+                    iterErrorPolygonCompact(iter, centerErr);
+                    return;
+                }
+                if (pointInsidePolygon(iter->_polygon, iter->_bboxes,
+                                       &center)) {
+                    // Set to next output
+                    iter->cell = cell;
+                    return;
+                }
             }
-            if (pointInsidePolygon(iter->_polygon, iter->_bboxes, &center)) {
-                // Set to next output
-                iter->cell = cell;
-                return;
+            if (mode == CONTAINMENT_OVERLAPPING) {
+                // For overlapping, we need to do a quick check to determine
+                // whether the polygon is wholly contained by the cell. We check
+                // the first polygon vertex, which if it is contained could also
+                // mean we simply intersect.
+                H3Index polygonCell;
+                H3Error polygonCellErr = H3_EXPORT(latLngToCell)(
+                    &(iter->_polygon->geoloop.verts[0]), cellRes, &polygonCell);
+                if (polygonCellErr != E_SUCCESS) {
+                    iterErrorPolygonCompact(iter, polygonCellErr);
+                    return;
+                }
+                if (polygonCell == cell) {
+                    // Set to next output
+                    iter->cell = cell;
+                    return;
+                }
+            }
+            if (mode == CONTAINMENT_FULL || mode == CONTAINMENT_OVERLAPPING) {
+                CellBoundary boundary;
+                H3Error boundaryErr =
+                    H3_EXPORT(cellToBoundary)(cell, &boundary);
+                if (boundaryErr != E_SUCCESS) {
+                    iterErrorPolygonCompact(iter, boundaryErr);
+                    return;
+                }
+                BBox bbox;
+                H3Error bboxErr = cellToBBox(cell, &bbox, false);
+                if (bboxErr) {
+                    iterErrorPolygonCompact(iter, bboxErr);
+                    return;
+                }
```

Comment:
This case may not be testable because presumably the boundary error above would not be E_SUCCESS? Perhaps there exists some index for which cellToBoundary returns 0 but cellToBbox returns non-0?

Reply:
Related question, do we have fuzzer question of these functions?

Reply:
Not sure if this is coverable, will switch to `NEVER` if I can't figure out an approach.

We don't have a fuzzer for `polygonToCellsExperimental`, but I can add one - that's probably best in a separate PR.

Reply:
I would be happy to help getting a fuzzer added. Agreed doing that in a separate PR would make the most sense.